torch/nn/parallel/distributed.py

```diff
@@ -633,27 +633,60 @@ def _register_comm_hook(self, state: object, hook: callable):
                              c10d reducer would call this hook and use the tensors returned
                              by the Future and copy grads to individual parameters.
 
+                             We also provide an API called ``get_future`` to retrieve a
+                             Future associated with the completion of c10d.ProcessGroupNCCL.work.
+                             Note that this API will return a toch._C.Future which is an
```

Comment:
I might miss-communicated in my previous comments. This `hook()` should return `torch.futures.Future`, as this is a requirement for users. The `get_future()` API returns `torch._C.Future`, as we should clarify how it is different from `torch.futures.Future` in the API doc of `get_future()`.

Think from a user's perspective. This API doc should tell a user how to write a hook. If we say the hook should return a `torch._C.Future`, then the user might be confused about what is `torch._C.Future`? how can I construct `torch._C.Future`? Why this should return `torch._C.Future`?

The reason we let `get_future()` return a `torch._C.Future` is not because `hook()` should return `torch._C.Future`. It is only a compromise we made for NCCL operations, and the `hook()` API is serving a broader range of applications than NCCL ops. So the limitation and special-casing should be on `get_future()` instead of on `hook()`.

Reply:
Referring to the thread in https://github.com/pytorch/pytorch/pull/42335#discussion_r465207732 

By saying "this API will return a toch._C.Future", I was meaning "get_future" API. Revised the explanation. Hope looks better now.